src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManager.cs

```diff
@@ -0,0 +1,499 @@
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public class VmrManager : IVmrManager
+{
+    private const string HEAD = "HEAD";
+
+    // Message shown when synchronizing a single commit
+    private const string SingleCommitMessage =
+        """
+        [{name}] Sync {newShaShort}: {commitMessage}
+
+        Original commit: {remote}/commit/{newSha}
+        """;
+
+    // Message shown when synchronizing multiple commits as one
+    private const string SquashCommitMessage =
+        """
+        [{name}] Sync {oldShaShort} → {newShaShort}
+        Diff: {remote}/compare/{oldSha}..{newSha}
+        
+        From: {remote}/commit/{oldSha}
+        To: {remote}/commit/{newSha}
+        
+        Commits:
+        {commitMessage}
+        """;
+
+    private readonly ILogger<VmrManager> _logger;
+    private readonly IProcessManager _processManager;
+    private readonly IRemoteFactory _remoteFactory;
+    private readonly string _vmrPath;
+    private readonly string _reposPath;
+    private readonly string _tagsPath;
+    private readonly string _tmpPath;
+
+    public IReadOnlyCollection<SourceMapping> Mappings { get; }
+
+    public VmrManager(
+        IProcessManager processManager,
+        IRemoteFactory remoteFactory,
+        ILogger<VmrManager> logger,
+        IReadOnlyCollection<SourceMapping> mappings,
+        string vmrPath,
+        string tmpPath)
+    {
+        _logger = logger;
+        _processManager = processManager;
+        _remoteFactory = remoteFactory;
+        _tmpPath = tmpPath;
+        _vmrPath = vmrPath;
+        _reposPath = Path.Join(vmrPath, "src");
+        _tagsPath = Path.Join(_reposPath, ".tags");
+
+        Mappings = mappings;
+    }
+
+    public async Task InitializeVmr(
+        SourceMapping mapping,
+        string? targetRevision,
+        bool ignoreWorkingTree,
+        CancellationToken cancellationToken)
+    {
+        if (File.Exists(Path.Combine(_reposPath, $".{mapping.Name}")))
+        {
+            throw new EmptySyncException($"Repository {mapping.Name} already exists");
+        }
+
+        _logger.LogInformation("Initializing {name}", mapping.Name);
+
+        string clonePath = await PrepareClone(mapping);
+        string patchPath = GetPatchFilePath(mapping);
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        using var clone = new Repository(clonePath);
+        var commit = GetCommit(clone, targetRevision is null || targetRevision == HEAD ? null : targetRevision);
+
+        await CreatePatch(mapping, clonePath, Constants.EmptyGitObject, commit.Id.Sha, patchPath);
+        cancellationToken.ThrowIfCancellationRequested();
+        await ApplyPatch(mapping, patchPath, cleanWorkingTree: !ignoreWorkingTree);
+        await TagRepo(mapping, commit.Id.Sha);
+
+        var description = $"[{mapping.Name}] Initial pull of the individual repository ({ShortenId(commit.Id.Sha)})" +
```

Comment:
Are we mixing string interpolation and addition for code formating?

Reply:
Right, we could use here the multiline string as well